core/src/main/java/org/apache/druid/indexer/partitions/PartitionsSpec.java

```diff
@@ -75,4 +76,29 @@ static Integer resolveHistoricalNullIfNeeded(@Nullable Integer val)
   {
     return isEffectivelyNull(val) ? null : val;
   }
+
+  /**
+   * @return True if this partitionSpec's type is compatible with forceGuaranteedRollup=true.
+   */
+  @JsonIgnore
+  default boolean isForceGuaranteedRollupCompatibleType()
+  {
+    return !(this instanceof DynamicPartitionsSpec);
+  }
+
+  /**
+   * @return True if this partitionSpec's property values are compatible with forceGuaranteedRollup=true.
+   */
+  @JsonIgnore
+  default boolean isForceGuaranteedRollupCompatible()
+  {
+    return getForceGuaranteedRollupIncompatiblityReason().isEmpty();
+  }
+
+  /**
+   * @return Message describing why this partitionSpec is incompatible with forceGuaranteedRollup=true. Empty string if
+   * the partitionSpec is compatible.
+   */
+  @JsonIgnore
+  String getForceGuaranteedRollupIncompatiblityReason();
```

Comment:
note to self: there are 3 ways of saying whether something is incompatible or not. Is there a better way to do this?